debugger/lldb-extension/lldb_sgx_plugin.py

```diff
@@ -0,0 +1,313 @@
+
+    OFFSETOF_MODULES = 80
+    SIZEOF_MODULES = 8
+
+    def __init__(self, addr):
+        if addr:
+            self.magic = read_int_from_memory(addr + self.OFFSETOF_MAGIC, self.SIZEOF_MAGIC)
+        if not self.is_valid():
+            return
+
+        self.version = read_int_from_memory(addr + self.OFFSETOF_VERSION, self.SIZEOF_VERSION)
+        self.next = read_int_from_memory(addr + self.OFFSETOF_NEXT, self.SIZEOF_NEXT)
+
+        path = read_int_from_memory(addr + self.OFFSETOF_PATH, self.SIZEOF_PATH)
+        path_length = read_int_from_memory(addr + self.OFFSETOF_PATH_LENGTH, self.SIZEOF_PATH_LENGTH)
+        self.path = bytes(read_from_memory(path, path_length)).decode('utf-8')
+
+        self.base_address = read_int_from_memory(addr + self.OFFSETOF_BASE_ADDRESS, self.SIZEOF_BASE_ADDRESS)
+
+        self.tcs = []
+        self.num_tcs = read_int_from_memory(addr + self.OFFSETOF_TCS_COUNT, self.SIZEOF_TCS_COUNT)
+        tcs_ptr = read_int_from_memory(addr + self.OFFSETOF_TCS_ARRAY, self.SIZEOF_TCS_ARRAY)
+        for i in range(0, self.num_tcs):
+            tcs = read_int_from_memory(tcs_ptr, 8) # sizeof pointer is hard-coded to 8
+            self.tcs.append(tcs)
+            tcs_ptr += 8
+
+        flags = read_int_from_memory(addr + self.OFFSETOF_FLAGS, self.SIZEOF_FLAGS)
+        self.debug = bool(flags & self.MASK_DEBUG)
+        self.simulate = bool(flags & self.MASK_SIMULATE)
+
+        self.modules = read_int_from_memory(addr + self.OFFSETOF_MODULES, self.SIZEOF_MODULES)
+
+    def is_valid(self):
+        return self.magic == self.MAGIC_VALUE
+
+
+# This constant definition must align with sgx_tcs_t
+TCS_GSBASE_OFFSET =  56
+
+# The set to store all loaded OE enclave base address.
+g_loaded_oe_enclave_addrs = set()
+
+# Global enclave list parsed flag
+g_enclave_list_parsed = False
+
+def read_from_memory(addr, size):
+    process = lldb.debugger.GetSelectedTarget().GetProcess()
+    """Read data with specified size  from the specified memory"""
+    # ( check the address is inside the enclave)
+    if addr == 0:
+        print ("Error happens in read_from_memory: addr = {0:x}".format(int(addr)))
+        return None
+
+    # process.ReadMemory() returns "memory read failed for" error while reading tcs_addr at least in lldb-7 and earlier
+    # that's why use system call to read memory
+    pid = process.GetProcessID()
+
+    fd = os.open("/proc/" + str(pid) + "/mem", os.O_RDONLY)
+    os.lseek(fd, int(addr), 0)
+    memory = os.read(fd, size)
+    os.close(fd)
+
+    if memory != -1:
+        return memory
+    else:
+        print ("Can't access memory at {0:x}.".format(int(addr)) + "\n" + str(os.error))
+        return None
+
+def read_int_from_memory(addr, size):
+    mv = read_from_memory(addr, size)
+    return int.from_bytes(mv, 'little')
+
+def target_path_to_host_path(target_path):
+    so_name = os.path.basename(target_path)
+    strpath = gdb.execute("show solib-search-path", False, True)
```

Comment:
Question: Are we supposed to use `gdb` here?

Reply:
Good catch. Will fix.

Reply:
Fixed. It is not clear what scenario target_path_to_host_path addresses. I have removed it for oelldb.